crates/cli/src/commands/build/mod.rs

```diff
@@ -0,0 +1,523 @@
+        NodePlatform::Windows => {
+          format!("{}.dll", cdylib)
+        }
+        _ => {
+          format!("lib{}.so", cdylib)
+        }
+      };
+
+      let dest_name = format!(
+        "{}{}.node",
+        "index",
+        if self.args.platform {
+          format!(".{}", self.target.platform_arch_abi)
+        } else {
+          "".to_owned()
+        }
+      );
+
+      Some((src_name, dest_name))
+    } else if let Some(bin) = &self.bin_target {
+      let src_name = if self.target.platform == NodePlatform::Windows {
+        format!("{}.exe", bin)
+      } else {
+        bin.clone()
+      };
+
+      let dest_name = src_name.clone();
+
+      Some((src_name, dest_name))
+    } else {
+      None
+    }
+  }
+
+  fn process_type_def(&self) {
+    if !self.intermediate_type_file.exists() {
+      return;
+    }
+
+    let mut dest = self.output_dir.clone();
+    match &self.args.dts {
+      Some(dts) => dest.push(dts),
+      None => dest.push("index.d.ts"),
+    };
+
+    let type_def_file = IntermidiateTypeDefFile::from(&self.intermediate_type_file);
+    let dts = type_def_file
+      .into_dts(!self.args.no_dts_header)
+      .expect("Failed to parse type def file");
+
+    write_file(&dest, &dts).expect("Failed to write type def file");
+  }
+
+  fn write_js_binding(&self) {
+    if !self.args.platform || self.args.disable_js_binding {
+      return;
+    }
+
+    let mut output = self.output_dir.clone();
+    output.push(
+      self
+        .args
+        .js_binding
+        .clone()
+        .unwrap_or_else(|| String::from("index.js")),
+    );
+
+    let mut env = Environment::new();
+    env
+      .add_template("index.js", include_str!("./templates/binding.tpl"))
+      .unwrap();
+
+    let binding = env
+      .get_template("index.js")
+      .and_then(|template| {
+        template.render(context!(
+          binary_name => self.package.name.clone(),
+          package_name => self.args.js_package_name.clone().unwrap_or_else(|| self.package.name.clone())
+        ))
+      })
+      .expect("Failed to generate js binding file.");
+
+    write_file(&output, &binding).expect("Failed to write js binding file");
+  }
+
+  fn use_zig(&self, cmd: &mut Command) -> &Self {
+    if !self.args.zig {
+      return self;
+    }
+
+    let mut zig_target = self.target.triple.clone();
+    if let Some(suffix) = &self.args.zig_abi_suffix {
+      zig_target.push('.');
+      zig_target.push_str(suffix);
+    }
+
+    let (cc, cxx) =
+      cargo_zigbuild::zig::prepare_zig_linker(&zig_target).expect("Failed to prepare zig linker");
+
+    let linker_env_name = format!(
+      "CARGO_TARGET_{}_LINKER",
+      self.target.triple.to_uppercase().replace('-', "_")
+    );
+
+    let envs = [("CC", &cc), ("CXX", &cxx), (linker_env_name.as_str(), &cc)];
+    trace!("set envs for zig cross compiling");
+    for (name, value) in envs.iter() {
+      trace!("  {}: {}", &name, &value.display());
+      cmd.env(name, value);
+    }
```

Comment:
The env vars are better aligned with `cargo-zigbuild`: https://github.com/messense/cargo-zigbuild/blob/9f31dcb97d5cff618d60b3c1d18dfb9f51045b7c/src/build.rs#L184-L233. 

And there are some hacks we do in `cargo-zigbuild` for some targets: https://github.com/messense/cargo-zigbuild/blob/9f31dcb97d5cff618d60b3c1d18dfb9f51045b7c/src/build.rs#L239-L287

Reply:
The complexity of using `cargo-zigbuild` is much more than my expectation.

I would migrate the old logic from js version which makes it work first and well design the crossing compiling part later on.

Reply:
Well, it's much easier if you do it this way: https://github.com/PyO3/maturin/blob/922459f19d5a1ba1295401f6b9c13198bcf61c2b/src/compile.rs#L202-L221 , all the env vars and operating system dependency hacks are taken care of.